nemo_skills/finetuning/data_preparation_utils/preprocessing.py

```diff
@@ -0,0 +1,222 @@
+        samples = []
+        if self.prediction_jsonl_files:
+            samples.extend(self._read_raw_data())
+        if self.preprocessed_dataset_files:
+            samples.extend(self._read_preprocessed_data())
+        with open(self.output_manifest_file, "wt", encoding="utf-8") as fout:
+            for sample in samples:
+                fout.write(json.dumps(sample) + "\n")
+
+
+class GroupSamples(BaseProcessor):
+    def __init__(self, group_key='question', **kwargs):
+        super().__init__(**kwargs)
+        self.group_key = group_key
+
+    def process(self):
+        samples = defaultdict(list)
+        with open(self.input_manifest_file, "rt", encoding="utf-8") as fin:
+            for line in fin:
+                sample = json.loads(line)
+                samples[sample[self.group_key]].append(sample)
+
+        with open(self.output_manifest_file, "wt", encoding="utf-8") as fout:
+            for groupped_samples in samples.values():
+                fout.write(json.dumps(groupped_samples) + "\n")
+
+
+class ShuffleAndDownsampleData(BaseProcessor):
+    def __init__(self, random_seed: int, num_samples: Optional[int] = None, sampling_method: str = 'random', **kwargs):
+        super().__init__(**kwargs)
+        self.sampling_method = sampling_method
+        self.num_samples = num_samples
+        self.random_seed = random_seed
+
+    def process(self):
+        groupped_samples = []
+        with open(self.input_manifest_file, "rt", encoding="utf-8") as fin:
+            for line in fin:
+                samples = json.loads(line)
+                groupped_samples.append(samples)
+
+        random.seed(self.random_seed)
+        if self.sampling_method == "random":
+            output_instances = list(chain(*groupped_samples))
+            random.shuffle(output_instances)
+            output_instances = output_instances[: self.num_samples]
+        elif self.sampling_method == "fair":
+            soln_counter = 0
+            output_instances = []
+            while True:
+                for quesn_idx in range(len(groupped_samples)):
+                    if len(output_instances) == self.num_samples:
+                        break
+                    if len(groupped_samples[quesn_idx]) > soln_counter:
+                        output_instances.append(groupped_samples[quesn_idx][soln_counter])
+                soln_counter += 1
+                if len(output_instances) == self.num_samples:
+                    break
+            random.shuffle(output_instances)
+        else:
+            raise NotImplementedError(f"Sampling method {self.sampling_method} not implemented")
+
+        with open(self.output_manifest_file, "wt", encoding="utf-8") as fout:
+            for instance in output_instances:
+                fout.write(json.dumps(instance) + "\n")
+
+
+class WriteFinalSftManifest(BaseProcessor):
+    def __init__(self, prompt_type: str, chat_format: bool = False, metadata: Dict = {}, **kwargs):
```

Comment:
Please don't use mutable defaults like `{}`, instead set to None and check inside if it's None, then create an empty dict